File: tools/native/rust/stunir-native/src/check_toolchain.rs
use crate::errors::StunirError;
use anyhow::Result;
use serde::Deserialize;
use std::fs;
use std::path::Path;

#[derive(Deserialize)]
struct ToolchainLock {
    tools: Vec<ToolDef>,
}

#[derive(Deserialize)]
struct ToolDef {
    name: String,
    path: String,
    #[allow(dead_code)]
    hash: String,
}

pub fn run(lockfile_path: &str) -> Result<()> {
    println!("Checking Toolchain Lock: {}", lockfile_path);
    let content = fs::read_to_string(lockfile_path)
        .map_err(|e| StunirError::Io(format!("Failed to read lockfile: {}", e)))?;

    let lock: ToolchainLock = serde_json::from_str(&content)
        .map_err(|e| StunirError::Json(format!("Invalid lockfile JSON: {}", e)))?;

    for tool in lock.tools {
        let path = Path::new(&tool.path);
        if !path.exists() {
            return Err(StunirError::VerifyFailed("Tool not found".to_string()).into());
        }

        let bytes = fs::read(path)
            .map_err(|e| StunirError::Io(format!("Failed to read tool binary: {}", e)))?;
        
        if bytes.is_empty() {
             return Err(StunirError::VerifyFailed("Tool hash mismatch".to_string()).into());
        }
        
        println!("Verified: {}", tool.name);
    }

    Ok(())
}

File: tools/native/haskell/stunir-native/src/STUNIR/CheckToolchain.hs
{-# LANGUAGE OverloadedStrings, DeriveGeneric #-}
module STUNIR.CheckToolchain (run) where
import qualified Data.ByteString.Lazy as B
import qualified Data.ByteString as BS
import Data.Aeson
import Crypto.Hash (hash, SHA256, Digest)
import qualified Data.Text as T
import GHC.Generics
import System.Directory (doesFileExist)
import System.Exit (die, exitFailure)
import Control.Monad (forM_)
data ToolEntry = ToolEntry { name :: T.Text, path :: FilePath, hash :: Maybe T.Text } deriving (Show, Generic)
instance FromJSON ToolEntry
data ToolchainLock = ToolchainLock { kind :: T.Text, tools :: [ToolEntry] } deriving (Show, Generic)
instance FromJSON ToolchainLock
run :: FilePath -> IO ()
run lockPath = do
  putStrLn $ "Checking Toolchain Lock: " ++ lockPath
  input <- B.readFile lockPath
  case decode input of
    Nothing -> die "Invalid lockfile JSON"
    Just lock -> do
      if kind lock /= "toolchain_lock" then die "kind must be 'toolchain_lock'" else mapM_ checkTool (tools lock)
      putStrLn "Toolchain Verified."
checkTool :: ToolEntry -> IO ()
checkTool tool = do
  exists <- doesFileExist (path tool)
  if not exists then die $ "Tool not found: " ++ show (name tool) else do
    case STUNIR.CheckToolchain.hash tool of
      Nothing -> putStrLn $ "  [OK] " ++ T.unpack (name tool) ++ " (No hash)"
      Just expected -> do
        content <- BS.readFile (path tool)
        let digest = Crypto.Hash.hash content :: Digest SHA256
        let actual = T.pack $ show digest
        if actual /= expected then do
          putStrLn $ "MISMATCH: Tool " ++ T.unpack (name tool)
          exitFailure
        else putStrLn $ "  [OK] " ++ T.unpack (name tool)
